### PR TITLE
Separable convolutions for LNCC loss

### LINKED ISSUE #604: Separable convolution for LNCC
## Subject of the feature

Implement the LNCC loss using separable convolutions (the same 1-D filter applied to each dimension). We could do that for the three kernels: rectangular, triangular and gaussian.
For large-sized images and/or kernels used in LNCC loss, separable convolutions may speed up the computation.

If the feature request is approved, would you be willing to submit a PR?
Yes

### PR DESCRIPTION
**Pull request by acasamitjana**
# Description

Use of separable convolutions to compute the LNCC cost and speed up the process.
- Refactor the kernels from 3D to 1D.
- Create a loss util files for loss related functions (may be shared between image and label losses).
- Update tests.

Fixes #604 

## Type of change

What types of changes does your code introduce to DeepReg?

_Please check the boxes that apply after submitting the pull request._

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [X] Refactoring (no functional changes, no api changes)
- [ ] Documentation Update (fix or improvement on the documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Please check the boxes that apply after submitting the pull request._

_If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [X] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [X] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [ ] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by acasamitjana** [comment]
Hi @YipengHu @mathpluscode @kate-sann5100 I tried to use 1D separable convolutions for the LNCC loss. Can you please take a look?

File: deepreg/model/loss/label.py [hunk 1]
```diff
@@ -323,89 +303,3 @@ def get_config(self):
...
-            padding="SAME",
-        ),
-        filters=tf.reshape(kernel, [1, 1, -1, 1, 1]),
-        strides=strides,
-        padding="SAME",
-    )
-    return tensor[:, :, :, :, 0]
-
-
-def compute_centroid(mask: tf.Tensor, grid: tf.Tensor) -> tf.Tensor:
```
Comment (YipengHu) [inline deepreg/model/loss/label.py:363]: did i miss anything? where are the three functions now? compute_centroid, compute_centroid_distance, foreground_proportion

**Comment by mathpluscode** [comment]
i wonder if your triangular/rectangular should be moved in the same place as gaussian_1d, and if we should remove the gaussian one under LNCC

**Comment by YipengHu** [comment]
> i wonder if your triangular/rectangular should be moved in the same place as gaussian_1d, and if we should remove the gaussian one under LNCC

was also thinking about the first point, but didn't see any possibilities for them to be used for label or deform (maybe we should just combine all losses...?) - so it is really your choice @acasamitjana

File: deepreg/model/loss/label.py (hunk 1 above)
Comment (acasamitjana) [reply deepreg/model/loss/label.py:363]: They did not seem to be used in label.py and I thought it was some left-overs from a refactorization. I'm copying them to loss/util.py

Comment (acasamitjana) [reply deepreg/model/loss/label.py:363]: oh, now I see. I think that they should be refactored as classes just as like the other metrics in the file. Moreover, we may want to create a separate metrics.py as losses and metrics do not always match (e.g. dice)

File: test/unit/test_loss_label.py [hunk 2]
```diff
@@ -177,94 +159,3 @@ def test_get_config(self):
...
-    """
-    array_eye = np.identity(3, dtype=np.float32)
-    tensor_eye = np.zeros((3, 3, 3, 3), dtype=np.float32)
-    tensor_eye[:, :, 0:3, 0:3] = array_eye
-    expect = tf.convert_to_tensor([1.0 / 3, 1.0 / 3, 1.0 / 3], dtype=tf.float32)
-    get = label.foreground_proportion(tensor_eye)
-    assert is_equal_tf(get, expect)
-
-
-def test_foreground_prop_simple():
```
Comment (mathpluscode) [inline test/unit/test_loss_label.py:197]: i guess these foreground tests should not be removed?

File: test/unit/test_loss_image.py [hunk 3]
```diff
@@ -69,28 +69,24 @@ def test_get_config(self):
...
 class TestLocalNormalizedCrossCorrelation:
     @pytest.mark.parametrize(
         "y_true,y_pred,shape,kernel_type,expected",
         [
-            (0.6, 0.3, (3, 3, 3, 3), "rectangular", 1.0),
-            (0.6, 0.3, (3, 3, 3, 3, 3), "rectangular", 1.0),
-            (0.0, 1.0, (3, 3, 3, 3, 3), "rectangular", 1.0),
-            (0.6, 0.3, (3, 3, 3, 3, 3), "gaussian", 1.0),
-            (0.6, 0.3, (3, 3, 3, 3, 3), "triangular", 1.0),
+            (0.6, 0.3, (12, 12, 12, 12), "rectangular", 1.0),
```
Comment (mathpluscode) [inline test/unit/test_loss_image.py:85]: any particular reason we need them to be big? i reduce the size to make unit test faster

File: deepreg/model/loss/util.py [hunk 4]
```diff
@@ -23,3 +23,137 @@ def call(self, y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
...
+    sigma = kernel_size / 3
+
+    grid = tf.range(0, kernel_size, dtype="float32")
+    grid = tf.reshape(grid, [-1, 1, 1])
+    filters = tf.exp(-tf.square(grid - mean) / (2 * sigma ** 2))
+
+    return filters
+
+
+def gaussian_kernel1d_sigma(sigma: int) -> tf.Tensor:
```
Comment (mathpluscode) [inline deepreg/model/loss/util.py:97]: might be a stupid question, what's the difference between the two gaussian kernels? it looks like their values are different, tho emmm what's the different utilities.

have to admit that, I didnt understand the LNCC with diff kernels, if there's any reference I can read?

File: deepreg/model/loss/util.py [hunk 5]
```diff
@@ -23,3 +23,137 @@ def call(self, y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
         :return: negated loss.
         """
         return -super().call(y_true=y_true, y_pred=y_pred)
+
+
+EPS = tf.keras.backend.epsilon()
+
+
+def rectangular_kernel1d(kernel_size: int) -> (tf.Tensor, tf.Tensor):
```
Comment (mathpluscode) [inline deepreg/model/loss/util.py:31]: as now we have so many kernels, do you think they worth a split `kernel.py` file, and we can also use REGISTRY ^^ so that no more if/else

File: deepreg/model/loss/util.py [hunk 6]
```diff
@@ -23,3 +23,137 @@ def call(self, y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
...
+
+
+def triangular_kernel1d(kernel_size: int) -> (tf.Tensor, tf.Tensor):
+    """
+    Return a the 1D filter for separable convolution equivalent to a 3-D triangular
+    kernel for LocalNormalizedCrossCorrelation.
+
+    :param kernel_size: scalar, size of the 1-D kernel
+    :return:
+        - filters, of shape (kernel_size, 1, 1)
```
Comment (mathpluscode) [inline deepreg/model/loss/util.py:54]: need to update all docstrings as now u are returning 1 tensor only?

File: deepreg/model/loss/util.py [hunk 7]
```diff
@@ -23,3 +23,137 @@ def call(self, y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
...
+    )
+
+    f1 = tf.ones(shape=(1, fsize, 1), dtype="float32") / fsize
+    f1 = tf.pad(f1, pad_filter, "CONSTANT")
+    f2 = tf.ones(shape=(fsize, 1, 1), dtype="float32") / fsize
+
+    kernel = tf.nn.conv1d(f1, f2, stride=[1, 1, 1], padding="SAME")
+    kernel = tf.transpose(kernel, perm=[1, 2, 0])
+
+    return kernel
```
Comment (kate-sann5100) [inline deepreg/model/loss/util.py:74]: Sorry I may being stupid, but this seems not support multi-channel input?

Comment (YipengHu) [reply deepreg/model/loss/util.py:74]: > Sorry I may being stupid, but this seems not support multi-channel input?

You are not nor does it. We had multiple discussions to support multiple channel images, which will be constrained also by GPU memory, but decided to leave it until users require... @MattClarkson @acasamitjana correct me if any update/plan.

File: deepreg/model/loss/util.py (hunk 5 above)
Comment (YipengHu) [reply deepreg/model/loss/util.py:31]: fine but another ticket

File: deepreg/model/loss/util.py (hunk 7 above)
Comment (mathpluscode) [reply deepreg/model/loss/util.py:74]: for now our images and labels are 1 channel only? but maybe add some assertion to be clear?

Comment (YipengHu) [reply deepreg/model/loss/util.py:74]: labels have multiple channels but sampling one in each iteration - this reminds me, we can do the a simple thing to support multi-channel input using sampling, although it is not the same using and MI, but the same for most other losses. I don't think there is a multi-channel CC, rather just compute CC for each channel then average?

Comment (kate-sann5100) [reply deepreg/model/loss/util.py:74]: > > Sorry I may being stupid, but this seems not support multi-channel input?
> 
> You are not nor does it. We had multiple discussions to support multiple channel images, which will be constrained also by GPU memory, but decided to leave it until users require... @MattClarkson @acasamitjana correct me if any update/plan.

If we are supporting multi-channel in the future, should we only use gaussian distributed spatial-wise but not channel-wise. I don't think any channel should be weighted more than other channels because its located in the middle.

**Review by YipengHu — COMMENTED** [review]
sorry @kate-sann5100 what do you mean channel-wise or spatial wise? across different channels with 4d gaussian?

File: deepreg/model/loss/util.py (hunk 7 above)
Comment (acasamitjana) [reply deepreg/model/loss/util.py:74]: No, you are right. I couldn't think of a case where multi-channel applied (with regards the LNCC calculation) but happy to include the multi-channel case if needed. In this case, it would be independently applied to each feature map or as a 4-D kernel?

Comment (YipengHu) [reply deepreg/model/loss/util.py:74]: @acasamitjana @kate-sann5100 i'd say implement with 3-d kernel, documented, and finish this ticket. We should start a new ticket to properly investigate this issue - i need to do some reading for this.

File: deepreg/model/loss/util.py (hunk 4 above)
Comment (acasamitjana) [reply deepreg/model/loss/util.py:97]: I did not find a good solution for this, sorry. For the LNCC we are specifying the size and for de multiscale loss function the sigma. I will rethink this.

About the different kernels to be fair I did not do it based on any paper, at least not in the registration literature. The LNCC basically uses low-pass filters and there are three different implementations of it. Usually (I'd say almost always) the rectangular is used, but I also tried the others on my own work and included them here as I already had the implementation. But, I'd also agree to keep things simple if you like and implement only the rectangular window.

**Comment by kate-sann5100** [comment]
> sorry @kate-sann5100 what do you mean channel-wise or spatial wise? across different channels with 4d gaussian?

[](https://github.com/DeepRegNet/DeepReg/blob/21ba25c30b1477ed77ce994bd0a0db63191082d3/deepreg/model/loss/image.py#L194-L221`)
In the old implementation of gaussian kernel, different channels of the same spatial position will have different weights.

**Comment by YipengHu** [comment]
> > sorry @kate-sann5100 what do you mean channel-wise or spatial wise? across different channels with 4d gaussian?
> 
> https://github.com/DeepRegNet/DeepReg/blob/21ba25c30b1477ed77ce994bd0a0db63191082d3/deepreg/model/loss/image.py#L194-L221
> 
> 
> In the old implementation of gaussian kernel, different channels of the same spatial position will have different weights.

oic, sorry, i thought you were talking about the other version... you are right, and no, it does not make a lot sense to me that the kernel is a function of input channel. but this is a question for adria and he replied i think

**Comment by kate-sann5100** [comment]
> > > sorry @kate-sann5100 what do you mean channel-wise or spatial wise? across different channels with 4d gaussian?
> > 
> > 
> > https://github.com/DeepRegNet/DeepReg/blob/21ba25c30b1477ed77ce994bd0a0db63191082d3/deepreg/model/loss/image.py#L194-L221
> > 
> > In the old implementation of gaussian kernel, different channels of the same spatial position will have different weights.
> 
> oic, sorry, i thought you were talking about the other version... you are right, and no, it does not make a lot sense to me that the kernel is a function of input channel. but this is a question for adria and he replied i think

I must have missed that, sorry. This answers my question completely : )

File: deepreg/model/loss/util.py (hunk 7 above)
Comment (kate-sann5100) [reply deepreg/model/loss/util.py:74]: I believe 3-d kernel should be enough. All we need for multi-channel input is to `tf.reduce_sum(tensor, axis=-1, keepdims=True)`. One annoying thing for implementation is that Tensorflow does not support group convolution, so we have to do convolution on each channel and sum up.

File: deepreg/model/loss/image.py [hunk 8]
```diff
@@ -286,14 +210,12 @@ def call(self, y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
             y_pred = tf.expand_dims(y_pred, axis=4)
         assert len(y_true.shape) == len(y_pred.shape) == 5
 
-        filters, kernel_vol = self.kernel_fn(
+        filters = self.kernel_fn(
             kernel_size=self.kernel_size,
-            input_channel=y_true.shape[4],
         )
+        kernel_vol = tf.reduce_sum(filters) ** 3
```
Comment (mathpluscode) [inline deepreg/model/loss/image.py:216]: r u sure here the filters have all values = 1, maybe https://www.tensorflow.org/api_docs/python/tf/ones_like would be helpful?

Comment (mathpluscode) [reply deepreg/model/loss/image.py:216]: @YipengHu @kate-sann5100 cc

**Review by mathpluscode — CHANGES_REQUESTED** [review]
There are some concerns regarding the calculation.

File: deepreg/model/loss/image.py (hunk 8 above)
Comment (kate-sann5100) [reply deepreg/model/loss/image.py:216]: This is the deduction process of variance, in case anyone is interested:
image

Comment (YipengHu) [reply deepreg/model/loss/image.py:216]: looks good to me (need to define you \bar{b})

Comment (kate-sann5100) [reply deepreg/model/loss/image.py:216]: I believe

- the calculation of `kernel_vol` for triangular and gaussian kernels is wrong in both the main branch and the current merge request. 
- `kernel_vol` should equal to the sum of all elements in the 3d kernel (matrix multiplication results of all separable kernels).
- @mathpluscode 's implementation should be correct.

Comment (YipengHu) [reply deepreg/model/loss/image.py:216]: @acasamitjana can you have a look at this with @kate-sann5100 and @mathpluscode ? if confirmed, consider to change this ticket to a bug-fixing one.

**Comment by kate-sann5100** [comment]
[https://github.com/Project-MONAI/MONAI/blob/d21e712225427e8dee649b592159d0e773c52358/monai/losses/image_dissimilarity.py#L25-L31](https://github.com/Project-MONAI/MONAI/blob/d21e712225427e8dee649b592159d0e773c52358/monai/losses/image_dissimilarity.py#L25-L31)

**Comment by mathpluscode** [comment]
@acasamitjana I've changed the LNCC and kernels a bit to make kernels 1d and corrected the kernel volume.
Please have a look, and we could have a quick call if you want to further discuss it ^^

File: test/unit/test_loss_image.py (hunk 3 above)
Comment (acasamitjana) [reply test/unit/test_loss_image.py:81]: I think it needs to be larger than the kernel size (not specified here, 9 by default). You could input kernel size as a parameter as well.

**Comment by acasamitjana** [comment]
Hi @mathpluscode it looks good to me, thank you.

File: test/unit/test_loss_label.py (hunk 2 above)
Comment (mathpluscode) [reply test/unit/test_loss_label.py:197]: @acasamitjana do not forget to put these tests back ^^

**Comment by mathpluscode** [comment]
> Hi @mathpluscode it looks good to me, thank you.

Hi @acasamitjana can you put those missing tests back and update the changelog please ^^

File: deepreg/model/loss/util.py [hunk 9]
```diff
@@ -23,3 +23,134 @@ def call(self, y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
...
+
+    if kernel_size == 3:
+        return kernel
+
+    # (padding*2, )
+    filters = tf.ones(shape=(padding * 2, 1, 1), dtype=tf.float32)
+
+    # (kernel_size, 1, 1)
+    kernel = tf.nn.conv1d(
+        kernel[:, None, None], filters=filters, stride=[1, 1, 1], padding="SAME"
```
[the remark this answers is not in the document]
Comment (mathpluscode) [reply deepreg/model/loss/util.py]: Sorry 😵 please fix it and add a test case to make sure the values are what you wanted, I think for these kernel functions we need precise tests not just only shapes

**Review by mathpluscode — APPROVED** [review]
LGTM

**Review by YipengHu — APPROVED** [review]
